api/webrtc/webrtc.go

```diff
@@ -0,0 +1,345 @@
+package webrtc
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/nknorg/nkn/v2/api/ratelimiter"
+	"github.com/nknorg/nkn/v2/api/websocket/session"
+	"github.com/nknorg/nkn/v2/config"
+	"github.com/nknorg/nkn/v2/util/log"
+	"github.com/pion/webrtc/v4"
+)
+
+// compitable to websocket
+const (
+	UnknownMessage = 0
+	TextMessage    = 1
+	BinaryMessage  = 2
+	CloseMessage   = 8
+	PingMessage    = 9
+	PongMessage    = 10
+
+	PingData = "ping"
+	PongData = "pong"
+)
+
+var zeroTime = time.Time{}
+var NewConnection func(conn session.Conn, r *http.Request)
+
+type DataChannelMessage struct {
+	messageType int
+	data        []byte
+}
+
+type Peer struct {
+	pc          *webrtc.PeerConnection
+	dc          *webrtc.DataChannel
+	offer       string
+	answer      string
+	OnSdp       chan string
+	OnConnected chan bool
+	OnMessage   chan DataChannelMessage
+
+	readDeadline  time.Time
+	writeDeadline time.Time
+	readLimit     int64
+	pongHandler   func(string) error
+}
+
+func NewPeer(urls []string) *Peer {
+	p := &Peer{
+		OnSdp:       make(chan string, 1),
+		OnConnected: make(chan bool, 1),
+		OnMessage:   make(chan DataChannelMessage, 128),
+	}
+
+	config := webrtc.Configuration{
+		ICEServers: []webrtc.ICEServer{
+			{
+				URLs: urls,
+			},
+		},
+	}
+	var err error
+	pc, err := webrtc.NewPeerConnection(config) // api.NewPeerConnection(config) //
+	if err != nil {
+		return nil
+	}
+
+	p.pc = pc
+	return p
+}
+
+func (c *Peer) Offer(lable string) error {
+	c.pc.OnICECandidate(func(candidate *webrtc.ICECandidate) {
+		if candidate == nil {
+			localDesc := c.pc.LocalDescription()
+			encodedDescr := Encode(localDesc)
+			c.offer = encodedDescr
+			c.OnSdp <- encodedDescr
+		}
+	})
+
+	dc, err := c.pc.CreateDataChannel(lable, nil)
+	if err != nil {
+		return err
+	}
+	dc.OnOpen(func() {
+		log.Debugf("data channel %v has been opened\n", dc.Label())
+		c.OnConnected <- true
```

Comment:
need to add default to prevent stuck since the chan is not read anywhere

Reply:
Yes, modified.